cpp/serve/grammar/grammar_impl.h

```diff
@@ -0,0 +1,170 @@
+/*!
+ *  Copyright (c) 2023 by Contributors
+ * \file serve/grammar/grammar_impl.h
+ * \brief The header for data structures for grammar-guided generation.
+ */
+
+#ifndef MLC_LLM_SERVE_GRAMMAR_GRAMMAR_IMPL_H_
+#define MLC_LLM_SERVE_GRAMMAR_GRAMMAR_IMPL_H_
+
+#include <tvm/runtime/container/string.h>
+#include <tvm/runtime/registry.h>
+
+#include "../encoding.h"
+#include "grammar.h"
+
+namespace mlc {
+namespace llm {
+namespace serve {
+
+using namespace tvm::runtime;
+
+/*!
+ * \brief Stores the abstract syntax tree of a BNF grammar, and handles the grammar.
+ * \details The AST contains two sets: a set of rules, and a set of subrules.
+ * Each rule represents a production of the BNF grammar. The BNF grammar is a set of rules.
+ *
+ * Subrule means a part of the rule definition. For example, in the following rule:
+ * rule ::= ("a" "b") | "c"
+ * ("a" "b"), "c", ("a" "b") | "c" are all subrules.
+ *
+ * Rule is a represented with subrule and a name. Each rule has a rule id for reference.
+ *
+ * Subrule has several types:
+ * - Character range: a range of characters (each character is a unicode codepoint),
+ *   e.g. [a-z], [ac-z]
+ * - Not character range: all characters that are not in the range, e.g. [^a-z], [^ac-z]
+ * - Empty: an empty string, i.e. ""
+ * - Rule reference: a reference to another rule
+ * - Sequence: a sequence of subrules, e.g. ("a" "b"). These subrules are concatenated together.
+ * - Or rule: a choice of subrules, e.g. ("a" "b") | "c". Each subrule can be matched.
+ *
+ * Every subrule is represented by a variable-length vector of TData (int32_t), where the first
+ * element indicates the type of the subrule, and the subsequent elements represent the data of the
+ * subrule. We store all subrules in one vector and record the starting position of each subrule. At
+ * the same time, each subrule corresponds to an subrule id for reference.
+ *
+ * For the format in every subrule, see BNFGrammarImpl::DataKind.
+ */
+class BNFGrammarImpl : public BNFGrammarNode {
+ public:
+  /*! \brief The id of a rule. Refers to the index of a rule in the rule vector. */
+  using TRuleId = int32_t;
+  /*! \brief The id of a subrule. */
+  using TSubruleId = int32_t;
+
+  /****************** Rule handling ******************/
+  /*! \brief A rule with name. */
+  struct Rule {
+    std::string name;
+    TSubruleId subrule;
+  };
+
+  /*! \brief Insert a rule and return the rule id. */
+  TRuleId InsertRule(const Rule& rule) {
+    rules_.push_back(rule);
+    return rules_.size() - 1;
+  }
```

Comment:
I am wondering when this function is invoked. Right now my read is that when parsing a grammar file, we
* first insert all subrules to the grammar object,
* and then marking the rules among the subrules by invoking this function.
Am I understanding correctly?

And right now this function is only invoked in parser (I also left a comment there). Will it be invoked in somewhere else in the future PRs?

Reply:
Sorry, I think the process is
- Inserting all rules with subrule id -1 (to get the rule id for all rules for reference)
- Insert subrules
- Update the subrule field of all rules

I am also considering these interfaces. It's only some of them may be used by another pass BNFGrammarSimplifier to eliminate left recursion and remove redundant subrules. But that is not finished currently. Maybe I will do a round of refactor after submitting that pass.